### ISSUE TITLE
Cross-compile for armv7-unknown-linux-gnueabihf target fails

### ISSUE THREAD

**Issue by chessnokov** [alexcrichton/backtrace-rs#52]
`error: failed to run custom build command for `backtrace-sys v0.1.10`
process didn't exit successfully: `/home/chessnokov/workspace/volter/target/release/build/backtrace-sys-3989d5c503a9238f/build-script-build` (exit code: 101)
--- stdout
OPT_LEVEL = Some("3")
TARGET = Some("armv7-unknown-linux-gnueabihf")
HOST = Some("x86_64-unknown-linux-gnu")
TARGET = Some("armv7-unknown-linux-gnueabihf")
TARGET = Some("armv7-unknown-linux-gnueabihf")
HOST = Some("x86_64-unknown-linux-gnu")
CC_armv7-unknown-linux-gnueabihf = None
CC_armv7_unknown_linux_gnueabihf = None
TARGET_CC = None
CC = Some("arm-linux-gnu-gcc")
TARGET = Some("armv7-unknown-linux-gnueabihf")
HOST = Some("x86_64-unknown-linux-gnu")
CFLAGS_armv7-unknown-linux-gnueabihf = None
CFLAGS_armv7_unknown_linux_gnueabihf = None
TARGET_CFLAGS = None
CFLAGS = None
PROFILE = Some("release")
running: "/home/chessnokov/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-sys-0.1.10/src/libbacktrace/configure" "--with-pic" "--disable-multilib" "--disable-shared" "--disable-host-shared" "--target=armv7-unknown-linux-gnueabihf" "--host=x86_64-unknown-linux-gnu"
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking target system type... armv7-unknown-linux-gnueabihf
checking for x86_64-unknown-linux-gnu-gcc... arm-linux-gnu-gcc
checking for C compiler default output file name... 

--- stderr
configure: WARNING: If you wanted to set the --build type, don't use --host.
    If a cross compiler is detected then cross compile mode will be used.
configure: error: in `/home/chessnokov/workspace/volter/target/armv7-unknown-linux-gnueabihf/release/build/backtrace-sys-c41044212322c9f2/out':
configure: error: C compiler cannot create executables
See `config.log' for more details.
thread 'main' panicked at 'failed with: exit code: 77', /home/chessnokov/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-sys-0.1.10/build.rs:134
note: Run with `RUST_BACKTRACE=1` for a backtrace.
`

**Comment by alexcrichton** [comment]
Thanks for the report! I *think* that https://github.com/alexcrichton/backtrace-rs/commit/f3eb14fadf0a10939dec347ad7f11eaeea42d1c6 may be related here, mind testing that out and seeing if it works?

**Comment by alexcrichton** [comment]
Alright! [Looks like](https://github.com/alexcrichton/backtrace-rs/commit/f3eb14fadf0a10939dec347ad7f11eaeea42d1c6#commitcomment-22259406) it's working so I'm going to close this and publish a new version.

**Comment by U007D** [comment]
I'm cross-compiling for this same target (`cargo build --target=armv7-unknown-linux-gnueabihf`) and am getting:
```
   Compiling backtrace-sys v0.1.16
error: failed to run custom build command for `backtrace-sys v0.1.16`
process didn't exit successfully: `/Users/bRad/Development/he/projects/ada/target/debug/build/backtrace-sys-739404c9ff39383a/build-script-build` (exit code: 101)
--- stdout
OPT_LEVEL = Some("0")
TARGET = Some("armv7-unknown-linux-gnueabihf")
HOST = Some("x86_64-apple-darwin")
TARGET = Some("armv7-unknown-linux-gnueabihf")
TARGET = Some("armv7-unknown-linux-gnueabihf")
HOST = Some("x86_64-apple-darwin")
CC_armv7-unknown-linux-gnueabihf = None
CC_armv7_unknown_linux_gnueabihf = None
TARGET_CC = None
CC = Some("/usr/local/bin/clang")
TARGET = Some("armv7-unknown-linux-gnueabihf")
HOST = Some("x86_64-apple-darwin")
CFLAGS_armv7-unknown-linux-gnueabihf = None
CFLAGS_armv7_unknown_linux_gnueabihf = None
TARGET_CFLAGS = None
CFLAGS = None
DEBUG = Some("true")
running: "sh" "/Users/bRad/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-sys-0.1.16/src/libbacktrace/configure" "--with-pic" "--disable-multilib" "--disable-shared" "--disable-host-shared" "--host=armv7-unknown-linux-gnueabihf" "--build=x86_64-apple-darwin"
checking build system type... x86_64-apple-darwin
checking host system type... armv7-unknown-linux-gnueabihf
checking target system type... armv7-unknown-linux-gnueabihf
checking for armv7-unknown-linux-gnueabihf-gcc... /usr/local/bin/clang
checking for C compiler default output file name... 

--- stderr
configure: error: in `/Users/bRad/Development/he/projects/ada/target/armv7-unknown-linux-gnueabihf/debug/build/backtrace-sys-5a84541daa3280b6/out':
configure: error: C compiler cannot create executables
See `config.log' for more details.
thread 'main' panicked at 'failed with: exit code: 77', /Users/bRad/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-sys-0.1.16/build.rs:176:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.


Process finished with exit code 101
```
Is this the same issue or am I doing something wrong?

rustc 1.25.0-nightly (9fd7da904 2018-01-25)

**Comment by alexcrichton** [comment]
@U007D it looks like you have `CC=/usr/local/bin/clang` but that may not be configured as a cross compiler?

**Comment by U007D** [comment]
I'll poke around--I don't recall changing my config in the past month or so, but yes, I'll see which compilers I have installed capable of building for armv7.  Thanks, @alexcrichton.

**Comment by U007D** [comment]
I'm now on `arm-none-eabi-gcc`, and am getting the same failure...  Here is my current config.log.  Any thoughts?

```
This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.

It was created by package-unused configure version-unused, which was
generated by GNU Autoconf 2.64.  Invocation command line was

  $ /Users/bRad/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-sys-0.1.16/src/libbacktrace/configure --with-pic --disable-multilib --disable-shared --disable-host-shared --host=armv7-unknown-linux-gnueabihf --build=x86_64-apple-darwin

## --------- ##
## Platform. ##
## --------- ##

hostname = Socrates.local
uname -m = x86_64
uname -r = 16.7.0
uname -s = Darwin
uname -v = Darwin Kernel Version 16.7.0: Wed Oct  4 00:17:00 PDT 2017; root:xnu-3789.71.6~1/RELEASE_X86_64

/usr/bin/uname -p = i386
/bin/uname -X     = unknown

/bin/arch              = unknown
/usr/bin/arch -k       = unknown
/usr/convex/getsysinfo = unknown
/usr/bin/hostinfo      = Mach kernel version:
	 Darwin Kernel Version 16.7.0: Wed Oct  4 00:17:00 PDT 2017; root:xnu-3789.71.6~1/RELEASE_X86_64
Kernel configured for up to 8 processors.
4 processors are physically available.
8 processors are logically available.
Processor type: i486 (Intel 80486)
Processors active: 0 1 2 3 4 5 6 7
Primary memory available: 16.00 gigabytes
Default processor set: 346 tasks, 1803 threads, 8 processors
Load average: 2.14, Mach factor: 5.85
/bin/machine           = unknown
/usr/bin/oslevel       = unknown
/bin/universe          = unknown

PATH: /Users/bRad/.cargo/bin
PATH: /usr/local/cuda/lib
PATH: /usr/local/cuda/extras/CUPTI/lib
PATH: /Developer/NVIDIA/CUDA-8.0/lib
PATH: /Users/bRad/.cargo/bin
PATH: /usr/local/bin
PATH: /usr/bin
PATH: /bin
PATH: /usr/sbin
PATH: /sbin
PATH: /usr/local/MacGPG2/bin


## ----------- ##
## Core tests. ##
## ----------- ##

configure:2524: checking build system type
configure:2538: result: x86_64-apple-darwin
configure:2558: checking host system type
configure:2571: result: armv7-unknown-linux-gnueabihf
configure:2591: checking target system type
configure:2604: result: armv7-unknown-linux-gnueabihf
configure:2641: checking for armv7-unknown-linux-gnueabihf-gcc
configure:2668: result: /usr/local/bin/arm-none-eabi-gcc
configure:2937: checking for C compiler version
configure:2946: /usr/local/bin/arm-none-eabi-gcc --version >&5
arm-none-eabi-gcc (GNU Tools for ARM Embedded Processors 6-2017-q2-update) 6.3.1 20170620 (release) [ARM/embedded-6-branch revision 249437]
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

configure:2957: $? = 0
configure:2946: /usr/local/bin/arm-none-eabi-gcc -v >&5
Using built-in specs.
COLLECT_GCC=/usr/local/bin/arm-none-eabi-gcc
COLLECT_LTO_WRAPPER=/usr/local/Caskroom/gcc-arm-embedded/6-2017-q2-update/gcc-arm-none-eabi-6-2017-q2-update/bin/../lib/gcc/arm-none-eabi/6.3.1/lto-wrapper
Target: arm-none-eabi
Configured with: /tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/src/gcc/configure --target=arm-none-eabi --prefix=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/install-native --libexecdir=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/install-native/lib --infodir=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/install-native/share/doc/gcc-arm-none-eabi/info --mandir=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/install-native/share/doc/gcc-arm-none-eabi/man --htmldir=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/install-native/share/doc/gcc-arm-none-eabi/html --pdfdir=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/install-native/share/doc/gcc-arm-none-eabi/pdf --enable-languages=c,c++ --enable-plugins --disable-decimal-float --disable-libffi --disable-libgomp --disable-libmudflap --disable-libquadmath --disable-libssp --disable-libstdcxx-pch --disable-nls --disable-shared --disable-threads --disable-tls --with-gnu-as --with-gnu-ld --with-newlib --with-headers=yes --with-python-dir=share/gcc-arm-none-eabi --with-sysroot=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/install-native/arm-none-eabi --build=x86_64-apple-darwin10 --host=x86_64-apple-darwin10 --with-gmp=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/build-native/host-libs/usr --with-mpfr=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/build-native/host-libs/usr --with-mpc=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/build-native/host-libs/usr --with-isl=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/build-native/host-libs/usr --with-libelf=/tmp/jenkins-GCC-6-build_build-toolchain-mac-262_20170621_1498074986/build-native/host-libs/usr --with-host-libstdcxx='-static-libgcc -Wl,-lstdc++ -lm' --with-pkgversion='GNU Tools for ARM Embedded Processors 6-2017-q2-update' --with-multilib-list=rmprofile
Thread model: single
gcc version 6.3.1 20170620 (release) [ARM/embedded-6-branch revision 249437] (GNU Tools for ARM Embedded Processors 6-2017-q2-update) 
configure:2957: $? = 0
configure:2946: /usr/local/bin/arm-none-eabi-gcc -V >&5
arm-none-eabi-gcc: error: unrecognized command line option '-V'
arm-none-eabi-gcc: fatal error: no input files
compilation terminated.
configure:2957: $? = 1
configure:2946: /usr/local/bin/arm-none-eabi-gcc -qversion >&5
arm-none-eabi-gcc: error: unrecognized command line option '-qversion'; did you mean '--version'?
arm-none-eabi-gcc: fatal error: no input files
compilation terminated.
configure:2957: $? = 1
configure:2977: checking for C compiler default output file name
configure:2999: /usr/local/bin/arm-none-eabi-gcc -O0 -ffunction-sections -fdata-sections -fPIC -g -march=armv7-a -Wall -Wextra   conftest.c  >&5
/usr/local/Caskroom/gcc-arm-embedded/6-2017-q2-update/gcc-arm-none-eabi-6-2017-q2-update/bin/../lib/gcc/arm-none-eabi/6.3.1/../../../../arm-none-eabi/lib/thumb/v7-ar/libg.a(lib_a-exit.o): In function `exit':
exit.c:(.text.exit+0x1c): undefined reference to `_exit'
collect2: error: ld returned 1 exit status
configure:3003: $? = 1
configure:3040: result: 
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "package-unused"
| #define PACKAGE_TARNAME "libbacktrace"
| #define PACKAGE_VERSION "version-unused"
| #define PACKAGE_STRING "package-unused version-unused"
| #define PACKAGE_BUGREPORT ""
| #define PACKAGE_URL ""
| /* end confdefs.h.  */
| 
| int
| main ()
| {
| 
|   ;
|   return 0;
| }
configure:3046: error: in `/Users/bRad/Development/he/projects/ada/target/armv7-unknown-linux-gnueabihf/debug/build/backtrace-sys-5a84541daa3280b6/out':
configure:3050: error: C compiler cannot create executables
See `config.log' for more details.

## ---------------- ##
## Cache variables. ##
## ---------------- ##

ac_cv_build=x86_64-apple-darwin
ac_cv_env_CC_set=set
ac_cv_env_CC_value=/usr/local/bin/arm-none-eabi-gcc
ac_cv_env_CFLAGS_set=set
ac_cv_env_CFLAGS_value='-O0 -ffunction-sections -fdata-sections -fPIC -g -march=armv7-a -Wall -Wextra'
ac_cv_env_CPPFLAGS_set=
ac_cv_env_CPPFLAGS_value=
ac_cv_env_CPP_set=
ac_cv_env_CPP_value=
ac_cv_env_LDFLAGS_set=
ac_cv_env_LDFLAGS_value=
ac_cv_env_LIBS_set=
ac_cv_env_LIBS_value=
ac_cv_env_build_alias_set=set
ac_cv_env_build_alias_value=x86_64-apple-darwin
ac_cv_env_host_alias_set=set
ac_cv_env_host_alias_value=armv7-unknown-linux-gnueabihf
ac_cv_env_target_alias_set=
ac_cv_env_target_alias_value=
ac_cv_host=armv7-unknown-linux-gnueabihf
ac_cv_prog_CC=/usr/local/bin/arm-none-eabi-gcc
ac_cv_target=armv7-unknown-linux-gnueabihf

## ----------------- ##
## Output variables. ##
## ----------------- ##

ACLOCAL=''
ALLOC_FILE=''
AMTAR=''
AR='/usr/local/bin/arm-none-eabi-ar'
AUTOCONF=''
AUTOHEADER=''
AUTOMAKE=''
AWK=''
BACKTRACE_FILE=''
BACKTRACE_SUPPORTED=''
BACKTRACE_SUPPORTS_DATA=''
BACKTRACE_SUPPORTS_THREADS=''
BACKTRACE_USES_MALLOC=''
CC='/usr/local/bin/arm-none-eabi-gcc'
CFLAGS='-O0 -ffunction-sections -fdata-sections -fPIC -g -march=armv7-a -Wall -Wextra'
CPP=''
CPPFLAGS=''
CYGPATH_W=''
DEFS=''
DSYMUTIL=''
DUMPBIN=''
ECHO_C='\c'
ECHO_N=''
ECHO_T=''
EGREP=''
EXEEXT=''
EXTRA_FLAGS=''
FGREP=''
FORMAT_FILE=''
GREP=''
INSTALL_DATA=''
INSTALL_PROGRAM=''
INSTALL_SCRIPT=''
INSTALL_STRIP_PROGRAM=''
LD=''
LDFLAGS=''
LIBOBJS=''
LIBS=''
LIBTOOL=''
LIPO=''
LN_S=''
LTLIBOBJS=''
MAINT=''
MAINTAINER_MODE_FALSE=''
MAINTAINER_MODE_TRUE=''
MAKEINFO=''
MKDIR_P=''
NATIVE_FALSE=''
NATIVE_TRUE=''
NM=''
NMEDIT=''
OBJDUMP=''
OBJEXT=''
OTOOL64=''
OTOOL=''
PACKAGE=''
PACKAGE_BUGREPORT=''
PACKAGE_NAME='package-unused'
PACKAGE_STRING='package-unused version-unused'
PACKAGE_TARNAME='libbacktrace'
PACKAGE_URL=''
PACKAGE_VERSION='version-unused'
PATH_SEPARATOR=':'
PIC_FLAG=''
RANLIB='/usr/local/bin/arm-none-eabi-ranlib'
SED=''
SET_MAKE=''
SHELL='/bin/sh'
STRIP=''
VERSION=''
VIEW_FILE=''
WARN_FLAGS=''
ac_ct_CC=''
ac_ct_DUMPBIN=''
am__EXEEXT_FALSE=''
am__EXEEXT_TRUE=''
am__isrc=''
am__leading_dot=''
am__tar=''
am__untar=''
bindir='${exec_prefix}/bin'
build='x86_64-apple-darwin'
build_alias='x86_64-apple-darwin'
build_cpu='x86_64'
build_os='darwin'
build_vendor='apple'
datadir='${datarootdir}'
datarootdir='${prefix}/share'
docdir='${datarootdir}/doc/${PACKAGE_TARNAME}'
dvidir='${docdir}'
exec_prefix='NONE'
host='armv7-unknown-linux-gnueabihf'
host_alias='armv7-unknown-linux-gnueabihf'
host_cpu='armv7'
host_os='linux-gnueabihf'
host_vendor='unknown'
htmldir='${docdir}'
includedir='${prefix}/include'
infodir='${datarootdir}/info'
install_sh=''
libdir='${exec_prefix}/lib'
libexecdir='${exec_prefix}/libexec'
libtool_VERSION=''
localedir='${datarootdir}/locale'
localstatedir='${prefix}/var'
mandir='${datarootdir}/man'
mkdir_p=''
multi_basedir=''
oldincludedir='/usr/include'
pdfdir='${docdir}'
prefix='NONE'
program_transform_name='s,x,x,'
psdir='${docdir}'
sbindir='${exec_prefix}/sbin'
sharedstatedir='${prefix}/com'
sysconfdir='${prefix}/etc'
target='armv7-unknown-linux-gnueabihf'
target_alias='armv7-unknown-linux-gnueabihf'
target_cpu='armv7'
target_os='linux-gnueabihf'
target_vendor='unknown'

## ----------- ##
## confdefs.h. ##
## ----------- ##

/* confdefs.h */
#define PACKAGE_NAME "package-unused"
#define PACKAGE_TARNAME "libbacktrace"
#define PACKAGE_VERSION "version-unused"
#define PACKAGE_STRING "package-unused version-unused"
#define PACKAGE_BUGREPORT ""
#define PACKAGE_URL ""

configure: exit 77
```

**Comment by U007D** [comment]
I did manage to get the cross compile to work on my Linux machine, so I'll assume there is indeed something wrong with my Mac config and will go from there.

**Comment by Michael-F-Bryan** [comment]
I'm getting the same errors where cross-compiling to a Raspberry Pi (`arm-unknown-linux-gnueabihf`) fails. Regardless of whether I set `CC` to `clang` or `arm-none-eabi-gcc`, autotools still says *"The C compiler cannot create executables"*.

@U007D, do you remember what you changed to get it working?

**Comment by U007D** [comment]
@Michael-F-Bryan Yes, so weird, eh?  I did go back and figured out what was wrong on my Mac.  Indeed, the installed C-compilers I have were not cross-compilers.  But, like you, I installed arm-none-eabi-gcc (from https://launchpad.net/gcc-arm-embedded) and pointing the script to it, still got the same error.  I was baffled by this.

I had an image of a version of arm gcc I'd built from source on a *case sensitive* (**not** optional--I tried!) HFS+ filesystem using [crosstool-ng](https://github.com/crosstool-ng/crosstool-ng).  I mounted the image and referencing that compiler, the build for backtrace-rs went smoothly from there.  It was [a bit of a PITA](https://github.com/crosstool-ng/crosstool-ng/issues/711) to build arm gcc, though (more so than building a native compiler), but given that the bug I ran into is fixed, and the tool has continued to evolve since last May, it should be smoother now.  This, ultimately, was how I solved this on my Mac.

With that said, I have to believe that the binaries available from https://launchpad.net/gcc-arm-embedded should work.  I am no expert, but TBH, I did not look deeply enough into why they didn't work for me.  I knew this was building for me a month ago, and stopped looking once I remembered to mount my locally-built gcc (which you are welcome to, if it helps).
